### PR TITLE
refactor(core): updated payments response with payment_method_id & payment_method_status

### PR DESCRIPTION
**Pull request by prajjwalkumar17**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [x] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

This change incorporates every payment's response with 2 additional fields namely:
 -> `payment_method_id`
 -> `payment_method_status`

### Additional Changes

- [x] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context

The change will provide more context about the `payment_method` being used to carry out the payment.

## How did you test it?

We will have `payment_method_id` and `payment_method_status` in every payment's response(except Create).
Sample curl:
### `Payment` done using token: 
```
curl --location 'http://127.0.0.1:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_eZdRVkfE8sbYE7JSzm2Tu0HAca0PHRY6u3JoSFBukHuxoYeXw7bgwr7Gy2DJn1kS' \
--data-raw '{
    "amount": 6540,
    "currency": "USD",
    "confirm": false,
    "capture_method": "automatic",
    "capture_on": "2022-09-10T10:11:12Z",
    "amount_to_capture": 6540,
    "customer_id": "cus_v6LovN3Zs9k8QQ2UYcbi",
    "business_country": "US",
    "business_label": "default",
    "email": "guest@example.com",
    "name": "John Doe",
    "phone": "999999999",
    "phone_country_code": "+1",
    "description": "Its my first payment request",
    "authentication_type": "no_three_ds",
    "return_url": "https://google.com",
    "payment_method": "card",
    "payment_method_type": "credit",
    
    "payment_method_data": {
        "card": {
            "card_number": "4111111111111111",
            "card_exp_month": "10",
            "card_exp_year": "40",
            "card_holder_name": "John",
            "card_cvc": "737"
        }
    },
    "shipping": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594430",
            "country_code": "+91"
        }
    },
    "billing": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        }
    },
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "metadata": {
        "udf1": "value1",
        "new_customer": "true",
        "login_date": "2019-09-10T10:11:12Z"
    }
}

'
``` 

`Payment's Confirm`
```
{
    "payment_token": "token_XXXXXXXXX",
    "payment_method": "card",
    "card_cvc": "123"
}
``` 

Response should include these two fields
```
   "payment_method_id": "pm_pzFcNslye3gZLxUlmDbg",
    "payment_method_status": null
```
 (This will be updated after @vspecky [PR](https://github.com/juspay/hyperswitch/pull/4012) linking issue [here](https://github.com/juspay/hyperswitch/issues/4020))

### `Payment` done using Payment Data & Customer Acceptance: 
```
curl --location 'http://127.0.0.1:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_eZdRVkfE8sbYE7JSzm2Tu0HAca0PHRY6u3JoSFBukHuxoYeXw7bgwr7Gy2DJn1kS' \
--data-raw '{
    "amount": 6540,
    "currency": "USD",
    "confirm": false,
    "capture_method": "automatic",
    "capture_on": "2022-09-10T10:11:12Z",
    "amount_to_capture": 6540,
    "customer_id": "cus_v6LovN3Zs9k8QQ2UYcbi",
    "business_country": "US",
    "business_label": "default",
    "email": "guest@example.com",
    "name": "John Doe",
    "phone": "999999999",
    "phone_country_code": "+1",
    "description": "Its my first payment request",
    "authentication_type": "no_three_ds",
    "return_url": "https://google.com",
    "payment_method": "card",
    "payment_method_type": "credit",
    
    "payment_method_data": {
        "card": {
            "card_number": "4111111111111111",
            "card_exp_month": "10",
            "card_exp_year": "40",
            "card_holder_name": "John",
            "card_cvc": "737"
        }
    },
    "shipping": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594430",
            "country_code": "+91"
        }
    },
    "billing": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        }
    },
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "metadata": {
        "udf1": "value1",
        "new_customer": "true",
        "login_date": "2019-09-10T10:11:12Z"
    }
}
'
``` 

`Payment's Confirm`
```
curl --location 'http://127.0.0.1:8080/payments/pay_GnDHbieYt7GBDhJj79j3/confirm' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_eZdRVkfE8sbYE7JSzm2Tu0HAca0PHRY6u3JoSFBukHuxoYeXw7bgwr7Gy2DJn1kS' \
--data '{
    "payment_method": "card",
        "payment_method_data": {
        "card": {
            "card_number": "5555 5555 5555 4444",
            "card_exp_month": "03",
            "card_exp_year": "2030",
            "card_holder_name": "joseph Doe",
            "card_cvc": "737"
        }
    } ,  
  "retry_action": "manual_retry",
        "customer_acceptance": {
        "acceptance_type": "offline",
        "accepted_at": "1963-05-03T04:07:52.723Z",
        "online": {
            "ip_address": "in sit",
            "user_agent": "amet irure esse"
        }
    }
}'
``` 

Response should include these two fields
```
    "payment_method_id": "pm_oDb4RlrcsTxFZ5n1v3OA",
    "payment_method_status": "active"
```

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/core/payments/flows/setup_mandate_flow.rs [hunk 1]
```diff
@@ -109,7 +109,8 @@ impl Feature<api::SetupMandate, types::SetupMandateRequestData> for types::Setup
                 key_store,
                 is_mandate,
             ))
-            .await?;
+            .await?
+            .0;
```
Comment (vspecky) [inline crates/router/src/core/payments/flows/setup_mandate_flow.rs:113]: We need to send back the payment_method_id and payment_method_status in Setup Mandate flow as well. Please ensure you set it accordingly.

File: crates/router/src/core/payment_methods.rs [hunk 2]
```diff
@@ -44,7 +44,10 @@ pub trait PaymentMethodRetrieve {
         payment_intent: &PaymentIntent,
         card_token_data: Option<&CardToken>,
         customer: &Option<domain::Customer>,
-    ) -> RouterResult<Option<(payments::PaymentMethodData, enums::PaymentMethod)>>;
+    ) -> (
+        RouterResult<Option<(payments::PaymentMethodData, enums::PaymentMethod)>>,
+        Option<String>,
+    );
```
Comment (vspecky) [inline crates/router/src/core/payment_methods.rs:50]: `RouterResult` should encapsulate the entire return type. You can also consider creating a struct for the output, since it's becoming complex. You can put the struct definition in `router::types::storage::payment_method`.

```suggestion
    ) -> RouterResult<(Option<(payments::PaymentMethodData, enums::PaymentMethod)>, Option<String>)>;
```

File: crates/router/src/core/payments/helpers.rs [hunk 3]
```diff
@@ -1687,25 +1688,27 @@ pub async fn make_pm_data<'a, F: Clone, R, Ctx: PaymentMethodRetrieve>(
...
+                payment_method_details.attach_printable("in 'make_pm_data'")?;
 
             Ok::<_, error_stack::Report<errors::ApiErrorResponse>>(
                 if let Some((payment_method_data, payment_method)) = payment_method_details {
                     payment_data.payment_attempt.payment_method = Some(payment_method);
-                    Some(payment_method_data)
+                    (Some(payment_method_data), pm_id)
                 } else {
-                    None
+                    (None, None)
```
Comment (vspecky) [inline crates/router/src/core/payments/helpers.rs:1711]:
```suggestion
                    (None, pm_id)
```

File: crates/router/src/core/payments/helpers.rs [hunk 4]
```diff
@@ -2964,7 +2967,7 @@ pub async fn get_merchant_connector_account(
 /// # Arguments
 ///
 /// * `router_data` - original router data
-/// * `request` - new request
+/// * `request` - new requestcore/helper
```
Comment (vspecky) [inline crates/router/src/core/payments/helpers.rs:2970]:
```suggestion
/// * `request` - new request core/helper
```

**Review by vspecky — CHANGES_REQUESTED** [review]

**Review by vspecky — DISMISSED** [review]

File: config/development.toml [hunk 5]
```diff
@@ -3,7 +3,7 @@ enabled = false
 
 [log.console]
 enabled = true
-level = "DEBUG"
+level = "WARN"
```
Comment (srujanchikke) [inline config/development.toml:6]: Please revert this changes

File: crates/router/src/types/storage/payment_method.rs [hunk 6]
```diff
@@ -19,6 +23,12 @@ pub struct CardTokenData {
     pub token: String,
 }
 
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+pub struct PaymentMethodDataWithId {
+    pub payment_method_data: Option<(payments::PaymentMethodData, enums::PaymentMethod)>,
+    pub payment_method_id: Option<String>,
+}
```
Comment (Narayanbhat166) [inline crates/router/src/types/storage/payment_method.rs:30]:
```suggestion
pub struct PaymentMethodDataWithId {
    pub payment_method_data: Option<payments::PaymentMethodData>,
    pub payment_method: Option<payments::PaymentMethod>,
    pub payment_method_id: Option<String>,
}
```

File: crates/router/src/core/payments.rs [hunk 7]
```diff
@@ -1993,7 +1997,7 @@ where
 {
     // On confirm is false and only router related
     let payment_data = if !is_operation_confirm(operation) {
-        let (_operation, payment_method_data) = operation
+        let (_operation, payment_method_data, pm_id) = operation
```
Comment (Narayanbhat166) [inline crates/router/src/core/payments.rs:2000]: This can be expanded using the [destructuring syntax.](https://doc.rust-lang.org/rust-by-example/flow_control/match/destructuring/destructure_structures.html)

**Comment by Narayanbhat166** [comment]
Hey @prajjwalkumar17 can you help me understand why `payment_method_status` is added to the `router_data`? Is this value used / updated / created by the connector module? Do we need to have it in router data? If we need not expose this value to the connectors, then it should be in payment data only

File: crates/router/src/core/payments/flows/authorize_flow.rs [hunk 8]
```diff
@@ -131,12 +134,10 @@ impl Feature<api::Authorize, types::PaymentsAuthorizeData> for types::PaymentsAu
                         self.request.payment_method_type,
                         &key_store,
                     ))
-                    .await;
+                    .await?;
```
[the remark this answers is not in the document]
Comment (vspecky) [reply crates/router/src/core/payments/flows/authorize_flow.rs:137]: @Narayanbhat166 Do we want to update the payment attempt twice, in a tokio task as well as the post update trackers?

File: crates/router/src/core/payments/flows/authorize_flow.rs [hunk 9]
```diff
@@ -117,26 +122,26 @@ impl Feature<api::Authorize, types::PaymentsAuthorizeData> for types::PaymentsAu
...
 
-                    if let Err(err) = result {
-                        logger::error!("Asynchronously saving card in locker failed : {:?}", err);
+                match pm {
+                    Ok((payment_method_id, payment_method_status)) => {
+                        resp.payment_method_id = payment_method_id.clone();
+                        resp.payment_method_status = payment_method_status;
                     }
-                });
+                    Err(_) => logger::error!("Save pm to locker failed"),
```
Comment (vspecky) [inline crates/router/src/core/payments/flows/authorize_flow.rs:143]: log the error

File: crates/router/src/core/payments/flows/authorize_flow.rs [hunk 10]
```diff
@@ -117,26 +122,26 @@ impl Feature<api::Authorize, types::PaymentsAuthorizeData> for types::PaymentsAu
...
-                        &state,
-                        &connector,
-                        response,
-                        &maybe_customer,
-                        &merchant_account,
-                        self.request.payment_method_type,
-                        &key_store,
-                    ))
-                    .await;
+                let pm = Box::pin(tokenization::save_payment_method(
```
Comment (vspecky) [inline crates/router/src/core/payments/flows/authorize_flow.rs:127]: Nit: name the variable something more appropriate.

**Review by vspecky — CHANGES_REQUESTED** [review]

**Review by vspecky — APPROVED** [review]

**Review by vspecky — DISMISSED** [review]

**Review by apoorvdixit88 — APPROVED** [review]
Dashboard specific changes are ok

**Review by Narayanbhat166 — DISMISSED** [review]

**Review by srujanchikke — DISMISSED** [review]
Reviewed connector code!

**Review by apoorvdixit88 — APPROVED** [review]

**Review by srujanchikke — APPROVED** [review]

**Review by vspecky — APPROVED** [review]

**Review by Narayanbhat166 — APPROVED** [review]
